### PR TITLE
Issue 930

### PR DESCRIPTION
**Pull request by RyanDJLee**
Integrate HUnit to project, and implement starter test cases for testing parsers in ReqParser.

### CODE REVIEW

File: app/Database/Requirement.hs [hunk 1]
```diff
@@ -14,35 +14,35 @@ module Database.Requirement
...
 import qualified Data.List as L
 
-data ProgramReq = PRGREQ String [Req]
+data ProgramReq = PRGREQ String [Req] deriving (Eq, Show)
 
 -- Returns a well formatted String representing a program requirement for specified program.
-instance Show ProgramReq where
-    show (PRGREQ program reqs) = "Program Requirements for " ++ program
-                               ++ ":\n" ++ L.intercalate "\n" (map show reqs)
+-- instance Show ProgramReq where
```
Comment (david-yz-liu) [inline app/Database/Requirement.hs]: Eh, I think it's actually better to remove `ProgramReq` and `CourseReq` altogether. They won't be necessary for the final use of this parser.

File: app/Database/Requirement.hs [hunk 2]
```diff
@@ -14,35 +14,35 @@ module Database.Requirement
...
-    show (OR reqs) =
-        case reqs of
-        [x] -> show x
-        otherwise -> "(" ++ (L.intercalate "/" $ map show reqs) ++ ")"
-    show (FROM fces reqs) =  fces ++ " FCE(s) from:\n" ++ show reqs
-    show (GRADE grade reqs) =  "(" ++ show reqs ++ " with a minimum grade of " ++ grade ++ "%)"
-    show (RAW text) = "(" ++ text ++ ")"
+data Req = J String | AND [Req] | OR [Req] | FROM String Req | GRADE String Req | RAW String deriving (Eq, Show)
+
+-- instance Show Req where
```
Comment (david-yz-liu) [inline app/Database/Requirement.hs]: You can just delete this part, rather than commenting it out.

File: app/ParserTests/ParserTests.hs [hunk 3]
```diff
@@ -0,0 +1,117 @@
+{-|
+Description: Test Course Requirement Parsers using HUnit Testing Framework.
```
Comment (david-yz-liu) [inline app/ParserTests/ParserTests.hs]: @PkHutch please advise about the right field to add to this docstring here ("module" or something).

File: app/WebParsing/ReqParser.hs [hunk 4]
```diff
@@ -180,7 +180,9 @@ parseReqs reqString =
         Right x -> x
         Left e -> J (show e)
 
-    -- [] look for more conrner cases where separators are in english between
-    --    complex reqs.
-    -- [] year (Done by Christine)
-    -- [] create Group Type
+    -- [] When using sepby, we are not returning a list of reqs.. we are returning
```
Comment (david-yz-liu) [inline app/WebParsing/ReqParser.hs]: I believe we resolved this in today's meeting.

File: stack.yaml [hunk 5]
```diff
@@ -17,7 +17,7 @@ flags: {}
 extra-package-dbs: []
 
 # Control whether we use the GHC we find on the path
-# system-ghc: true
+system-ghc: false
```
Comment (david-yz-liu) [inline stack.yaml]: Please don't commit this change. (It's fine to have it on your local machine, but does not belong in this pull request.)

File: app/ParserTests/ParserTests.hs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import WebParsing.ReqParser
+import qualified Text.Parsec as Parsec
+import Text.Parsec.String (Parser)
+import Control.Monad.Identity (Identity)
+
+-- create function that takes in a parser and string,
+-- applies parser to string and creates assertion as in test 2?
+
+strTest1 :: Test
+strTest1 = TestCase (assertEqual "for (fcesParser \"5.0 FCEs\")," (Right "5.0") (Parsec.parse fcesParser "" "5.0 FCEs"))
```
Comment (david-yz-liu) [inline app/ParserTests/ParserTests.hs]: You should use your test-making helper for these tests.

File: app/ParserTests/ParserTests.hs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+--
+-- For more examples, check out the tests directory.  It contains unit tests
+-- for HUnit.
+
+module ParserTests.ParserTests
+( runStrTests,
+  runReqTests
+) where
+
+import Test.HUnit
```
Comment (david-yz-liu) [inline app/ParserTests/ParserTests.hs]: Specify explicitly which names you're importing from these libraries.

File: app/ParserTests/ParserTests.hs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+strTest10 = TestCase(let courseReq = Parsec.parse andParser "" " csc148h1,csc165h1 "
+                    in assertEqual "for (andParser \" csc148h1,csc165h1 \")," (Right $ AND[J"csc148h1",J"csc165h1"]) courseReq)
+
+strTest11 :: Test
+strTest11 = TestCase(let courseReq = Parsec.parse andParser "" " csc108h1/csc148h1,csc165h1 "
+                    in assertEqual "for (andParser \" csc108h1,csc148h1,csc165h1 \")," (Right $ AND[OR[J"csc108h1",J"csc148h1"],J"csc165h1"]) courseReq)
+
+-- Function to facilitate test case creation given a string, Req tuple
+createTest :: (String, Req) -> Parser Req -> Test
+createTest (input, expected) parser = TestCase (let courseReq =  Parsec.parse parser "" input
```
Comment (david-yz-liu) [inline app/ParserTests/ParserTests.hs]: It's more readable to have the `let` be the top-level construct, so that the `TestCase` constructor appears in the "in" part.

File: app/ParserTests/ParserTests.hs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    in assertEqual ("for (" ++ input ++ "),") (Right expected) courseReq)
+
+-- coBefTests
+coBefInputs :: [(String, Req)]
+coBefInputs = [(" minimum of 60% in csc148h1", GRADE "60" (J "csc148h1")),("a minimum of A in csc165h1", GRADE "A" (J "csc165h1")), (" a minimum of (B-) in csc165h1", GRADE "B-" (J "csc165h1"))]
```
Comment (david-yz-liu) [inline app/ParserTests/ParserTests.hs]: There are a few formatting issues on this line and in general to fix.

1. Start each element of a list on a new line, so that your lines aren't too long.
2. Fix whitespace using standard rules. Put a space after every comma, and after every value constructor (like `J`).

File: app/ParserTests/ParserTests.hs [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+coBefTests :: Test
+coBefTests = TestList $ (map (\parserToTest -> parserToTest coBefParser) (map createTest coBefInputs))
```
Comment (david-yz-liu) [inline app/ParserTests/ParserTests.hs]: See the discussion on Slack - you can merge the two `map` calls into one.

**Comment by RyanDJLee** [comment]
Addressed all changes. Need to get started on developing test suites.

File: app/ParserTests/ParserTests.hs (hunk 7 above)
Comment (david-yz-liu) [inline app/ParserTests/ParserTests.hs]: @RyanDJLee I think you missed this one.

Comment (RyanDJLee) [inline app/ParserTests/ParserTests.hs]: @david-yz-liu I forgot to mention this but I made a judgement call to leave out the names as I found that I had to specify ~ 6. I realized I could use Test(..) instead of explicitly importing each value constructor and reduced it 4. Sorry for not communicating this, I pushed the changes as requested.

**Comment by RyanDJLee** [comment]
Please refer to issue-930A for addressed changes.
